devito/passes/iet/misc.py

```diff
@@ -75,48 +73,58 @@ def relax_incr_dimensions(iet, **kwargs):
         if root in mapper:
             continue
 
+        # Split iterations to outer and inner
         outer, inner = split(iterations, lambda i: not i.dim.parent.is_Incr)
 
-        # Compute the iteration ranges
-        ranges = []
+        # Get symbolic_max out of each outer dimension
+        ranges = {}
         for i in outer:
-            maxb = i.symbolic_max - (i.symbolic_size % i.dim.step)
-            ranges.append(((i.symbolic_min, maxb, i.dim.step),
-                           (maxb + 1, i.symbolic_max, i.symbolic_max - maxb)))
-
-        # Remove any offsets
-        # E.g., `x = x_m + 2 to x_M - 2` --> `x = x_m to x_M`
-        outer = [i._rebuild(limits=(i.dim.root.symbolic_min, i.dim.root.symbolic_max,
-                                    i.step))
-                 for i in outer]
-
-        # Create the ElementalFunction
-        name = sregistry.make_name(prefix="bf")
-        body = compose_nodes(outer)
-        dynamic_parameters = flatten((i.symbolic_bounds, i.step) for i in outer)
-        dynamic_parameters.extend([i.step for i in inner if not is_integer(i.step)])
-        efunc = make_efunc(name, body, dynamic_parameters)
-
-        efuncs.append(efunc)
-
-        # Create the ElementalCalls
-        calls = []
-        for p in product(*ranges):
-            dynamic_args_mapper = {}
-            for i, (m, M, b) in zip(outer, p):
-                dynamic_args_mapper[i.symbolic_min] = m
-                dynamic_args_mapper[i.symbolic_max] = M
-                dynamic_args_mapper[i.step] = b
-                for j in inner:
-                    if j.dim.root is i.dim.root and not is_integer(j.step):
-                        value = j.step if b is i.step else b
-                        dynamic_args_mapper[j.step] = (value,)
-            calls.append(efunc.make_call(dynamic_args_mapper))
-
-        mapper[root] = List(body=calls)
-
-    iet = Transformer(mapper).visit(iet)
+            ranges[i.dim.root] = i.symbolic_max
+
+        # A dictionary to map maximum of parent dimensions
+        # useful for hierarchical blocking
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+
+            # Candidate 1: Retrieve parent's max from outer limits
+            parent_max = ranges[i.dim.root]
+
+            # Candidate 2: The symbolic_size of an inner Iteration
+            # may exceed the size of a parent's
+            # block size. Proper margin should be allowed
+            # For cases where upperbound - lowerbound > block_size:
+            # i.symbolic_size > i.dim.parent.step
+            low_margin = i.symbolic_min - i.dim.symbolic_min
+            size_extent = i.symbolic_size - i.dim.parent.step
```

Comment:
we never use the word "extent" for integer sets like iteration points / spaces. We use `extent` for physical quantities only, such at the resolution, in meters, of the grid. Etc etc